File: mobile/library/kotlin/src/io/envoyproxy/envoymobile/stats/GaugeImpl.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyEngine
import java.lang.ref.WeakReference

/**
 * Envoy implementation of a `Gauge`.
 */
internal class GaugeImpl : Gauge {
  internal val envoyEngine: WeakReference<EnvoyEngine>
  internal val elements: List<Element>

  internal constructor(engine: EnvoyEngine, elements: List<Element>) {
    this.envoyEngine = WeakReference<EnvoyEngine>(engine)
    this.elements = elements
  }

  override fun set(value: Int) {
    envoyEngine.get()?.recordGaugeSet(
      elements.joinToString(separator = ".") { it.element }, value
    )
  }

  override fun add(amount: Int) {
    envoyEngine.get()?.recordGaugeAdd(
      elements.joinToString(separator = ".") { it.element }, amount
    )
  }

  override fun sub(amount: Int) {
    envoyEngine.get()?.recordGaugeSub(
      elements.joinToString(separator = ".") { it.element }, amount
    )
  }
}

File: mobile/library/swift/src/stats/GaugeImpl.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// The implementation of time series gauge.
@objcMembers
class GaugeImpl: NSObject, Gauge {
  private let series: String
  private weak var engine: EnvoyEngine?

  init(elements: [Element], engine: EnvoyEngine) {
    self.series = elements.map { $0.value }.joined(separator: ".")
    self.engine = engine
    super.init()
  }

  /// Set the gauge with the given value.
  /// TODO: potentially raise error to platform if the operation is not successful.
  func set(value: Int) {
    guard let engine = self.engine else {
      return
    }

    engine.recordGaugeSet(self.series, value: numericCast(value))
  }

  /// Add the given amount to the gauge.
  /// TODO: potentially raise error to platform if the operation is not successful.
  func add(amount: Int) {
    guard let engine = self.engine else {
      return
    }

    engine.recordGaugeAdd(self.series, amount: numericCast(amount))
  }

  /// Subtract the given amount from the gauge.
  /// TODO: potentially raise error to platform if the operation is not successful.
  func sub(amount: Int) {
    guard let engine = self.engine else {
      return
    }

    engine.recordGaugeSub(self.series, amount: numericCast(amount))
  }
}
